completions.go

```diff
@@ -148,16 +166,21 @@ func (d ShellCompDirective) string() string {
 // Adds a special hidden command that can be used to request custom completions.
 func (c *Command) initCompleteCmd(args []string) {
 	completeCmd := &Command{
-		Use:                   fmt.Sprintf("%s [command-line]", ShellCompRequestCmd),
-		Aliases:               []string{ShellCompNoDescRequestCmd},
-		DisableFlagsInUseLine: true,
-		Hidden:                true,
-		DisableFlagParsing:    true,
-		Args:                  MinimumNArgs(1),
-		Short:                 "Request shell completion choices for the specified command-line",
+		Use:                fmt.Sprintf("%s [command-line]", ShellCompRequestCmd),
+		Aliases:            []string{ShellCompNoDescRequestCmd},
+		Hidden:             true,
+		DisableFlagParsing: true,
+		Args:               MinimumNArgs(1),
+		Short:              "Request shell completion choices for the specified command-line",
 		Long: fmt.Sprintf("%[2]s is a special command that is used by the shell completion logic\n%[1]s",
 			"to request completion choices for the specified command-line.", ShellCompRequestCmd),
 		Run: func(cmd *Command, args []string) {
+			args, _ = cmd.extractCompleteCmdFlags(args)
```

Comment:
I wonder if it is really necessary to have this implemented as flag. I think it would be much easier to just use a new cmd alias, .e.g. `ShellCompActiveHelpRequestCmd`. I am also not sure why `ActiveHelpConfig` should be a string, I personally think a boolean would be enough. Do you think people would implement several help levels?  I personally either want help or not.

Reply:
Thank you for having had a look already!

You bring up good points, all of which evolved as I tested Active Help.

> I think it would be much easier to just use a new cmd alias, e.g. `ShellCompActiveHelpRequestCmd`

You are right.  The problem happens when you want mix it with descriptions.  If I want descriptions but not active help, or vice versa, or both together, or neither one.  We would need a different command alias for every permutation.  And in the future, if we were to add a third feature, it would compound the problem.

An alternative to using a flag would be to use an environment variable.  That is actually how I originally implemented things.  However, when writing go tests for the program using Cobra, it became apparent that dealing with environment variables was not as easy as dealing with flags. But that point is open for discussion.

A better implementation for the flag solution would be to let Cobra parse the flag.  This could be done by separating the flag of the `__complete` command from the rest of the flags using the `--` delimiter (e.g., `helm __complete --__activfeHelpCfg=1 -- --namespace ''`).  The problem is that it wouldn't be backwards-compatible, as it would require programs to add the `--` to go tests they already have written; and the `--` would also become mandatory when debugging from the command-line.  That is why I passed on that approach.

> I am also not sure why ActiveHelpConfig should be a string, I personally think a boolean would be enough.

Another good point 😄 .  That was also how I first implemented things.  But when using the feature with helm, I hit some scenarios where the Active Help needed to query the kubernetes cluster; such queries were slow (in comparison).  This made me think that some programs may wish to have different levels of active help, e.g., off, on, local.  Since it is hard to predict the need of programs, I went with the full flexibility of a string.

Note however that a program can still choose to present a boolean flag to their users to only support active help being on or off.

Reply:
> You are right. The problem happens when you want mix it with descriptions. If I want descriptions but not active help, or vice versa, or both together, or neither one. We would need a different command alias for every permutation. And in the future, if we were to add a third feature, it would compound the problem.

Yep, that is what I thought.

> An alternative to using a flag would be to use an environment variable. That is actually how I originally implemented things. However, when writing go tests for the program using Cobra, it became apparent that dealing with environment variables was not as easy as dealing with flags. But that point is open for discussion.

I like the idea of environment variables because that will allow users to change the behaviour more easily. I find it tedious to regenerate the completion script, especially since for Podman we automatically install the script via package manager. Using environment variables would make it easy to document this. One global environment variable would allow me to change it for all my cobra applications.

As far as testing is concerned you can do `os.Setenv()` and `defer os.Unsetenv()` inside your go test function.

> A better implementation for the flag solution would be to let Cobra parse the flag. This could be done by separating the flag of the __complete command from the rest of the flags using the -- delimiter (e.g., helm __complete --__activfeHelpCfg=1 -- --namespace ''). The problem is that it wouldn't be backwards-compatible, as it would require programs to add the -- to go tests they already have written; and the -- would also become mandatory when debugging from the command-line. That is why I passed on that approach.

I don't think cobra needs to do the the flag parsing. We can use pflag directly, e.g. something like this should work:
https://github.com/containers/podman/blob/bbd03df3e120998f57393c25c14ca4fe15f5aec3/pkg/systemd/generate/containers.go#L268-L278

Can you provide a link to your helm changes. I would like to see how this looks in a real application.

Reply:
I've pushed a commit that switches to using an env var to continue the discussion.

Reply:
And you were right @Luap99 that using environment variables within go tests wasn't a problem.

But if we prefer using environment variables for active help, I don't think we should use only a global one as I mentioned in a previous comment.

I'll update the implementation to illustrate.